pkg/crdschema/gen.go

```diff
@@ -0,0 +1,377 @@
+		versionSchemas := map[string]*apiextensionsv1beta1.JSONSchemaProps{}
+		for pkg, gv := range p.GroupVersions {
+			if gv.Group != groupKind.Group {
+				continue
+			}
+			if !existingVersions[gv.Version] {
+				fmt.Printf("Source code for %s in %s/%s not found.\n", groupKind.Kind, gv.Group, gv.Version)
+				continue
+			}
+			versionSchemas[gv.Version] = nil
+
+			// find parser schema
+			typeIdent := crdgen.TypeIdent{Package: pkg, Name: groupKind.Kind}
+			typeInfo := p.Types[typeIdent]
+			if typeInfo == nil {
+				continue
+			}
+
+			// generate flattened schema
+			p.NeedSchemaFor(typeIdent)
+			versionSchemas[gv.Version] = crdgen.FlattenEmbedded(f.FlattenType(typeIdent), pkg)
+		}
+
+		// edge cases: no schema or all nil
+		allNil := true
+		for _, s := range versionSchemas {
+			if s != nil {
+				allNil = false
+				break
+			}
+		}
+		if allNil {
+			fmt.Printf("No schema found for %s in %s, skipping.\n", groupKind.Kind, groupKind.Group)
+			continue
+		}
+
+		// use global schema if there is just one, or all are equal
+		allEqual := true
+		var first *apiextensionsv1beta1.JSONSchemaProps
+		for _, s := range versionSchemas {
+			first = s
+			break
+		}
+		for _, s := range versionSchemas {
+			if !reflect.DeepEqual(s, first) {
+				allEqual = false
+				break
+			}
+		}
+
+		// delete old and set new schemas
+		var changed bool
+		if allEqual {
+			changed, err = existing.setGlobalSchema(first)
+			if err != nil {
+				return fmt.Errorf("failed to set global schema for %s in %s: %v", groupKind.Kind, groupKind.Group, err)
+			}
+		} else {
+			changed, err = existing.setVersionedSchema(versionSchemas)
+			if err != nil {
+				return fmt.Errorf("failed to set versioned schemas for %s in %s: %v", groupKind.Kind, groupKind.Group, err)
+			}
+		}
+		if changed {
+			fmt.Printf("Updating schema for %s in %s at spec.validation.openAPIV3Schema.\n", groupKind.Kind, groupKind.Group)
+		} else {
+			fmt.Printf("Unchanged schema for %s in %s at spec.validation.openAPIV3Schema.\n", groupKind.Kind, groupKind.Group)
+		}
+
+		// check whether the file changes
+		if changed, err := existing.compareToExistingFile(ctx, fn); err != nil {
+			return fmt.Errorf("failed to compare against %q: %v", fn, err)
+		} else if changed {
+			changedFiles = append(changedFiles, fn)
+		}
+	}
+
+	// only check for differences in verify-only mode
+	if g.VerifyOnly {
+		if len(changedFiles) > 0 {
+			return fmt.Errorf("unexpected changes to: %s", strings.Join(changedFiles, ", "))
+		}
+		return nil
+	}
+
+	return g.writeChangedFiles(ctx, changedFiles, existingCRDsByFileNames)
+}
+
+func (g Generator) writeChangedFiles(ctx *genall.GenerationContext, changedFiles []string, existingCRDsByFileNames map[string]*existingCRD) (result error) {
+	// write changes
+	for _, fn := range changedFiles {
+		bs, err := yaml.Marshal(existingCRDsByFileNames[fn].Yaml)
+		if err != nil {
+			return err
+		}
+		if ctx.OutputRule == genall.OutputToNothing {
```

Comment:
This isn't ok -- you're breaking the abstraction pretty dramatically here.

Reply:
done